### PR TITLE
Add some rebar3 / dialyzer related build stuff

### PR DESCRIPTION
**Pull request by sargun**

### CODE REVIEW

**Comment by sargun** [comment]
I'm about 75% done with this PR -- Do you guys have any interest in incorporating these fixes?

**Comment by sargun** [comment]
I realize that there are issues with rebar3 looking at the other PR. No reason we can't use rebar3 for dialyzer, and rebar2 for the real builds.

Also, bumping 17.3->17.5 is because SSL is broken af with 17.1, and 17.3

File: src/erldns_packet_cache.erl [hunk 1]
```diff
@@ -75,14 +77,16 @@ get(Question, _Host) ->
 %% @doc Put the response in the cache for the given question.
 -spec put(dns:question(), dns:message()) -> ok.
 put(Question, Response) ->
-  case ?ENABLED of
-    true ->
+  ?IF(
```
Comment (aeden) [inline src/erldns_packet_cache.erl]: Is this change necessary for some reason related to rebar3 or dialyzer?

File: src/erldns_query_throttle.erl [hunk 2]
```diff
@@ -103,6 +106,7 @@ code_change(_PreviousVersion, State, _Extra) ->
 
 
 % Internal API
+-spec(maybe_throttle(inet:ip_address() | inet:hostname()) -> throttle_result()).
```
Comment (aeden) [inline src/erldns_query_throttle.erl]: Is the throttle_result() type defined somewhere and I missed it?

File: src/erldns_resolver.erl [hunk 3]
```diff
@@ -148,7 +148,7 @@ resolve_exact_type_match(Message, Qname, Qtype, Host, CnameChain, MatchedRecords
             false ->
               resolve_exact_type_match(Message, Qname, Qtype, Host, CnameChain, MatchedRecords, Zone, _AuthorityRecords, NSRecords)
           end;
-        {error, authority_not_found} ->
+        {error, _} ->
```
Comment (aeden) [inline src/erldns_resolver.erl]: I do not believe this should be chained to allow all error types to be handled this way, only the case where the authority is missing.

**Comment by aeden** [comment]
I left some comments inline.

I'll be happy to incorporate fixes, as long as travis passes, the change set is clean and specific to the problem at hand, and that it doesn't break our production environment. 😉

File: src/erldns_packet_cache.erl (hunk 1 above)
Comment (sargun) [inline src/erldns_packet_cache.erl]: The reason for this is that when you guys are using macros to enable / disable things at compile time it makes a statement like
```
case ?ENABLED of
    true ->
        foo;
    false ->
        bar
end.

into:
case true of
    true ->
        foo;
    false ->
        bar
end.
```

Dialyzer doesn't quite understand this. This is basically an implementation of if-than-else with the parse transform implemented elsewhere.

File: src/erldns_query_throttle.erl (hunk 2 above)
Comment (sargun) [inline src/erldns_query_throttle.erl]: Yeah. Y'all already defined it in master: `https://github.com/aetrion/erl-dns/blob/master/src/erldns_query_throttle.erl#L39`

File: src/erldns_resolver.erl (hunk 3 above)
Comment (sargun) [inline src/erldns_resolver.erl]: Dialyzer doesn't like us explicitly wanting to crash. The other (theoretical) error case, based on the typespec:
```
 151: The pattern {'error', 'authority_not_found'} can never match the type {'error','no_question'} | {'ok',[any()]}
```
How should we handle `{error, no_question}`?

**Comment by sargun** [comment]
One of the comments that seems to have gotten lost is the issue with `erldns_resolver`. I had changed the error case here: https://github.com/aetrion/erl-dns/blob/master/src/erldns_resolver.erl#L143-L153 to `{error, _}`. The problem is that `erldns_zone_cache:get_authority` has the typespec to return `{'error','no_question'}` as well.

What do you propose we do? How should that error case be handled?

**Comment by sargun** [comment]
@aeden So, the build passes on Erlang 18, not on 17. Do y'all run erl-dns on 17? (If so, you really should rock 18.3, it's a free performance boost). I'd like to remove Erlang 17 from the builds. In fact, we've been running erl-dns on every `[DCOS](http://dcos.io)` node in Erlang 18.1, and it's been working.

**Comment by sargun** [comment]
(Also, a review would be chill)

**Comment by sargun** [comment]
Also, probably a good idea to delete `ebin/erldns.app` from git.

**Comment by aeden** [comment]
> Do y'all run erl-dns on 17?

We do. I am going to talk with other team members about moving to 18 in production.

File: src/erldns_resolver.erl (hunk 3 above)
Comment (aeden) [inline src/erldns_resolver.erl]: I understand now why you made the change here. Does it make sense to be explicit with a new case pattern for `{error, no_question}` that has the same logic as `{error, authority_not_found}` for now?

File: src/erldns_txt.erl [hunk 4]
```diff
@@ -77,15 +78,91 @@ split(Data, Parts) ->
...
 parse_test() ->
-  ?assertEqual(parse(""), []),
-  ?assertEqual(parse("test"), [[<<"test">>]]),
-  ?assertEqual(parse(lists:duplicate(270, "x")), [[list_to_binary(lists:duplicate(255, "x")), list_to_binary(lists:duplicate(15, "x"))]]),
-  ?assertEqual(parse(<<"test">>), [[<<"test">>]]),
-  ?assertEqual(parse("\"test\" \"test\""), [[<<"test">>], [<<"test">>]]),
-  ?assertEqual(parse("\\"), [[<<"\\">>]]),
-  ?assertEqual(parse("test\\;"), [[<<"test\\;">>]]),
-  ?assertEqual(parse("test\\"), [[<<"test\\">>]]).
+  ?assertEqual([], parse("")),
```
Comment (sargun) [inline src/erldns_txt.erl]: EUNIT expects ?assertEqual(Expected, Value). These were breaking, and resulting in confusion on my part.

File: src/erldns_txt.erl [hunk 5]
```diff
@@ -77,15 +78,91 @@ split(Data, Parts) ->
...
+  ?assertEqual([[<<"test">>]], parse("test")),
+  ?assertEqual([[list_to_binary(lists:duplicate(255, "x")), list_to_binary(lists:duplicate(15, "x"))]], parse(lists:duplicate(270, "x"))),
+  ?assertEqual([[<<"test">>]], parse(<<"test">>)),
+  ?assertEqual([[<<"test">>], [<<"test">>]], parse("\"test\" \"test\"")),
+  ?assertEqual([[<<"\\">>]], parse("\\")),
+  ?assertEqual([[<<"test\\;">>]], parse("test\\;")),
+  ?assertEqual([[<<"test\\">>]], parse("test\\")).
 %?assertEqual(parse("\"test\"\""), [[<<"test\"">>]]).
 
+proper_test_() ->
```
Comment (sargun) [inline src/erldns_txt.erl]: Added some model checks. We can get rid of these, or break them out into a separate commit, but it's was a really nice to have while troubleshooting this module.

**Comment by sargun** [comment]
So, it looks like it already wont work under Erlang 17 anyone since you're plugged into the new time API: https://github.com/aetrion/erl-dns/blob/master/src/erldns_storage.erl#L68-L70

**Comment by aeden** [comment]
I'm starting to review this now. I will let you know if I have questions.

**Comment by aeden** [comment]
Do you have a command or set of commands I should use to get dialyzer results?

File: .travis.yml [hunk 6]
```diff
@@ -1,3 +1,14 @@
 language: erlang
 otp_release:
-   - 17.3
+  - 18.1
```
Comment (aeden) [inline .travis.yml]: Want to bump this to 18.3?

File: src/erldns_resolver.erl [hunk 7]
```diff
@@ -148,6 +159,9 @@ resolve_exact_type_match(Message, Qname, Qtype, Host, CnameChain, MatchedRecords
             false ->
               resolve_exact_type_match(Message, Qname, Qtype, Host, CnameChain, MatchedRecords, Zone, _AuthorityRecords, NSRecords)
           end;
+        %%% There is an error condition here that dialyzer doesn't like -
+        %%% {error, no_question}
+        %%% we should / need to deal with it
```
Comment (aeden) [inline src/erldns_resolver.erl]: Go ahead with the wildcard match on Reason for now.

File: .travis.yml (hunk 6 above)
Comment (sargun) [inline .travis.yml]: Travis only supports up to 18.2.1: https://docs.travis-ci.com/user/languages/erlang -- I can bump it to that.

File: src/erldns_resolver.erl (hunk 7 above)
Comment (sargun) [inline src/erldns_resolver.erl]: After adding enough typespecs, it seems like dialyzer is happy with this.

File: .travis.yml (hunk 6 above)
Comment (aeden) [inline .travis.yml]: Interestingly Travis still built even when I specified 18.3 in one of our apps:

```
18.3 is not installed. Downloading and installing pre-build binary.
wget -o $HOME/erlang.tar.bz2 ${archive_url}
mkdir -p ~/otp && tar -xf erlang-18.3-nonroot.tar.bz2 -C ~/otp/
```

I wonder what it *actually* built with?

**Comment by sargun** [comment]
`./rebar3 dialyzer` should get you the results. It's part of the Travis build as well to check the typespecs.

**Comment by sargun** [comment]
I don't know the actual commands it runs under the hood. You can probably get them by doing `DEBUG=1 ./rebar3 dialyzer`, @aeden

**Comment by aeden** [comment]
Thanks for all the work on this, I appreciate it!
